### ISSUE TITLE
Timeouts

### ISSUE THREAD

**Issue by mattpawson** [Mimetis/Dotmim.Sync#1185]
Hi,

We have a very large database that I'm trying to create a sync for but it keeps timing out.

I have tried creating a snapshot for this but that times out as well.

We are using a direct connection to the database, as opposed to a web proxy.

Is there any way to increase this timeout / stop it from timing out?

Alternatively, is it possible to stop all connections to the server database, create the tracking data (triggers, tables, etc), and then backup the database and restore it as the client before re-enabling all connections to the database. That way when the sync is run, only the data that has been updated since the backup would be synced?

**Comment by Mimetis** [comment]
Hello @mattpawson 

I think your first sync, where everything is downloaded to the client, and prepared by the server, is where you have these timeouts.

Your idea to make a backup - restore on the client may be a good idea, indeed.

> **TL;DR:** Sample is here : https://github.com/Mimetis/Dotmim.Sync/tree/master/Samples/UseBackupOnClient

It's a little bit complicated, because you need to know how **DMS** is working behind the hood, but definitively something you can do.

Here are the "macro" tasks needed to achieve this:

- Backup your server database, without any **DMS** related metadata / tables / triggers / stored procedures. In my opinion, if you are doing the backup before **DMS** has provisioned the server database, it's easier. Otherwise choose carefully the tables you want to backup, and do not add the **DMS** metadata in the backup.
- Once you have backup your server database, you need to enable **DMS** to track all the data that are going to change from now, to be sure you will not loose any data added between the server backup and the client backup, using the `ProvisionAsync` method.
- You need also to get the timestamp from the server, that you will save to the client. It basically means "Here is the last sync timestamp from the server. Don't try to sync anything else before this timestamp". Using the `GetLocalTimestampAsync` method.
- Now you can restore the backup to your client.
- Then using the server scope (using `GetScopeInfoAsync` method), you can `ProvisionAsync` the client. 
- Client is ready, but does not know what is the starting time to sync from. This is where you are using the server timestamp and save it locally using the `GetScopeInfoClientAsync` method, then the `SaveScopeInfoClientAsync` method.

Here is the macro code:

``` csharp

// creating the orchestrators we are going to use
var serverProvider = new SqlSyncProvider(advConnectionString);
var clientProvider = new SqlSyncProvider(clientConnectionString);
var remoteOrchestrator = new RemoteOrchestrator(serverProvider);
var localOrchestrator = new LocalOrchestrator(clientProvider);

// Assuming adventureworks is already created, with data, without any DMS metadata

// backup AdventureWorks
Helper.BackupDatabase("AdventureWorks", advConnectionString);

// once we have the backup, we need to ensure all the data that will be added / edited / deleted on the server will be tracked
// we can do this by provision the server database
await remoteOrchestrator.ProvisionAsync(setup);

// get the correct timestamp from remote orchestrator.
// this timestamp will be recorded on the client, as a "from now on" timestamp
var serverTimestamp = await remoteOrchestrator.GetLocalTimestampAsync();

// from now on, we know all the future modification on the server are tracked
// As a test example, we are going to add a random line in ProductCategory :
await Helper.AddProductCategoryRowAsync(advConnectionString);

// restore AdventureWorks on the client, with a new name, called "Client"
Helper.RestoreSqlDatabase("AdventureWorks", "Client", masterConnectionString);

// get the server scope to be able to create the client scope
var serverScope = await remoteOrchestrator.GetScopeInfoAsync();

// provisionning the client, with all tracking tables and all DMS stuff
await localOrchestrator.ProvisionAsync(serverScope);

// get client timestamp that is different from the server timestamp
var clientTimestamp = await localOrchestrator.GetLocalTimestampAsync();

// get the client scope to update
var clientScope = await localOrchestrator.GetScopeInfoClientAsync();

// Override with correct values, to ensure next time we are running from the good starting point

// server last timestamp sync
clientScope.LastServerSyncTimestamp = serverTimestamp; 
// client last timestamp sync
clientScope.LastSyncTimestamp = clientTimestamp; 
// since we are considering the client as a restored database with existing values, we are not new anymore
clientScope.IsNewScope = false; 
// just as human readable value, not used internally
clientScope.LastSync = DateTime.UtcNow; 

await localOrchestrator.SaveScopeInfoClientAsync(clientScope);

```

From now on, you can make a classic Sync:

``` csharp
var advConnectionString = string.Format(sqlConnectionString, "AdventureWorks");
var clientConnectionString = string.Format(sqlConnectionString, "Client");

var serverProvider = new SqlSyncProvider(advConnectionString);
var clientProvider = new SqlSyncProvider(clientConnectionString);

// Creating an agent that will handle all the process
var agent = new SyncAgent(clientProvider, serverProvider);


do
{
    // Launch the sync process
    var s1 = await agent.SynchronizeAsync(setup);

    // Write results
    Console.WriteLine(s1);

} while (Console.ReadKey().Key != ConsoleKey.Escape);

Console.WriteLine("End");

```

Let me know if that helps

**Comment by Mimetis** [comment]
you're welcome dude ;) 
let's see if it's working for u !
